File: kotlin/beer-song/src/main/kotlin/BeerSong.kt
object BeerSong {
    private const val w = "on the wall"

    fun verses(startBottles: Int, takeDown: Int) =
        startBottles.downTo(takeDown).joinToString("\n") {
            val b = bottles(it)
            "$b $w, $b.\n".capitalize() +
            "${action(it)}, ${bottles(it - 1)} $w.\n"
        }

    private fun bottles(n: Int): String {
        val num = mapOf(-1 to "99", 0 to "no more")[n] ?: n.toString()
        return "$num bottle${if (n == 1) "" else "s"} of beer"
    }

    private fun action(n: Int): String = when (n) {
        0 -> "Go to the store and buy some more"
        else -> "Take ${if (n == 1) "it" else "one"} down and pass it around"
    }
}

File: fsharp/beer-song/BeerSong.fs
module BeerSong

(* Taking (excessive) care to not repeat any strings. *)

let private bottles num =
    let n = if num = 0 then "No more" else string num
    let s = if num = 1 then "" else "s"
    $"{n} bottle{s} of beer"

let private where = "on the wall"

let private firstLine (n: int): string =
    let b = bottles n
    $"{b} {where}, {b.ToLower()}."

let private secondLine (n: int): string =
    let one num = if num = 1 then "it" else "one"
    let b, action =
        match n with
        | 0 -> ((bottles 99), "Go to the store and buy some more")
        | _ -> ((bottles (n-1)), $"Take {one n} down and pass it around")
    $"{action}, {b.ToLower()} {where}." 

let recite (startBottles: int) (takeDown: int) = 
    [startBottles .. -1 .. (startBottles - takeDown + 1)]
    |> List.map (fun i -> [""; (firstLine i); (secondLine i)])
    |> List.concat
    |> List.skip 1